From 02bacf436f5e417f44963542b7e7b970db9f2f30 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Daniel=20Garc=C3=ADa=20Moreno?= <dani@danigm.net>
Date: Thu, 27 Jan 2022 16:39:20 +0100
Subject: [PATCH] API: Remove duplicates for tag filtering

Add distinct to the tag filter query.

The filter `tags__tag_name__in` is doing two INNER JOIN in the query
when there are more than one tag to filter by. This is producing
duplicate results.

```
SELECT * FROM "content_contentnode"
INNER JOIN "content_contentnode_tags" ON ("content_contentnode"."id" = "content_contentnode_tags"."contentnode_id")
INNER JOIN "content_contenttag" ON ("content_contentnode_tags"."contenttag_id" = "content_contenttag"."id")
WHERE (
    "content_contentnode"."available" = True AND
    "content_contentnode"."channel_id" = 057f871caa405ec29d62ba0523c193d7 AND
    "content_contenttag"."tag_name"
    IN (level=all levels, level=beginner, type=class)
) ORDER BY "content_contentnode"."tree_id" ASC, "content_contentnode"."lft" ASC
```

The added distinct filter fixes the issue.
---
 kolibri/core/content/api.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kolibri/core/content/api.py b/kolibri/core/content/api.py
index 4751381d1..a9bfb7272 100644
--- a/kolibri/core/content/api.py
+++ b/kolibri/core/content/api.py
@@ -299,7 +299,7 @@ class ContentNodeFilter(IdFilter):
         :return: content nodes that match the tags
         """
         tags = value.split(",")
-        return queryset.filter(tags__tag_name__in=tags).order_by("lft")
+        return queryset.filter(tags__tag_name__in=tags).order_by("lft").distinct()
 
     def filter_descendant_of(self, queryset, name, value):
         """
-- 
2.32.0

